Tmedia 324 lerna versioning independent

JackHowa opened the pull request:
## Description
Version blocks independently per branch/tag

## Jira Ticket
- [TMEDIA-324](https://arcpublishing.atlassian.net/secure/RapidBoard.jspa?rapidView=875&modal=detail&selectedIssue=TMEDIA-324&assignee=5e387d6a1b1d910e5dfd7a9b)

## Acceptance Criteria
only publish with lerna versions that have changed 

## Test Steps
- see demo and lerna-playground https://github.com/WPMedia/themes-lerna-playground

## Effect Of Changes
### Before
- all blocks published each time 
- slow and potentially dangerous with hitting rate limits 
- also long deploys mean that clients could publish mid-deploy with half the blocks updated

### After
- only changed blocks published 

## Dependencies or Side Effects
- Individual blocks track based on changes not release cycle 
- added secret 
<img width="892" alt="Screen Shot 2021-07-21 at 16 07 35" src="https://user-images.githubusercontent.com/5950956/126560249-c22425f7-6348-4243-9cc4-fb9cd1d1b8ce.png">


## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [ ] Confirmed all the test steps a reviewer will follow above are working.
- [ ] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [ ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ ] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
feat: Use beta and canary release and github builds

Commit:
ci: Add stable and rc github action for publishing

Commit:
ci: Only use the js and jsx paths for github actions

Commit:
ci: Update lerna publish targets, remove components publish check

Commit:
ci: Update lerna

Commit:
ci: Update lerna

Commit:
docs: Add lerna and github action info

Commit:
Merge branch 'canary' into TMEDIA-324-lerna-versioning-independent

Commit:
Update typo

Commit:
ci: Update package-lock based on merge

matthewroach commented on .github/workflows/beta-build.yml:
```diff
@@ -3,46 +3,61 @@ on:
+    # only run on changes to these files
+    paths:
+      - "**.js"
+      - "**.jsx"
```
Blocks can have `.scss` files which we would want to pick up on if the only change was to CSS

JackHowa replied:
ah good point, will fix

JackHowa replied:
updated

matthewroach commented on .github/workflows/beta-build.yml:
```diff
@@ -3,46 +3,61 @@ on:
       SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
     steps:
+      - name: Set git username
+        run: git config --global user.email "jack.howard@washpost.com" && git config --global user.name "Jack Howard"
```
We shouldn't really rely on specific developers accounts here, I've created a ticket to address this more globally - https://arcpublishing.atlassian.net/browse/TMEDIA-382

JackHowa replied:
that's a good point. @RobinGiannattasio talked about using a recently-created bot account that I think engine team is using, right? @badmintonkid

badmintonkid replied:
you'll need a user with publishing access. engine's is a dummy that doesn't have access to anything just so we can access public github packages. but you can do the same process to set one up. luke created a email distribution that we used for the email on the account signup.

JackHowa replied:
ah right thanks @badmintonkid

matthewroach commented on .github/workflows/canary-build.yml:
```diff
@@ -3,45 +3,59 @@ on:
+    # only run on changes to these files
+    paths:
+      - "**.js"
+      - "**.jsx"
```
Same as above, we should check for `.scss` files too

JackHowa replied:
updated

matthewroach commented on .github/workflows/rc-build.yml:
```diff
@@ -3,46 +3,56 @@ on:
+    # only run on changes to these files
+    paths:
+      - "**.js"
+      - "**.jsx"
```
Check for `.scss` files too

JackHowa replied:
updated

matthewroach commented on .github/workflows/stable-build.yml:
```diff
@@ -0,0 +1,66 @@
+    # only run on changes to these files
+    paths:
+      - "**.js"
+      - "**.jsx"
```
Check for `.scss` files too

matthewroach commented on .github/workflows/stable-build.yml:
```diff
@@ -0,0 +1,66 @@
+name: Stable blocks publish to stable tag
```
This is a new file for stable. Is the plan to rely on this automation for production releases going forward?

JackHowa replied:
I'd hope so. at least based on the ticket requirements

matthewroach commented on News Theme Development.md:
```diff
@@ -129,6 +129,36 @@ Common phrases needed across multiple blocks are declared in global-phrases-bloc
+
+- Only publish packages to beta
+- Publish packages based on only changes
+- Iterates the prerelease version using `--canary` arg that allows prerelease (not necessarily canary)
```
Shouldn't this read `--beta` ?

JackHowa replied:
--canary is (confusingly) a parameter for lerna https://github.com/lerna/lerna/tree/main/commands/publish#--canary

tried to explain that but guess it didn't come across. what would you recommend for an update

matthewroach commented on News Theme Development.md:
```diff
@@ -129,6 +129,36 @@ Common phrases needed across multiple blocks are declared in global-phrases-bloc
+
+- Only publish packages to rc
+- Publish packages based on only changes
+- Iterates the prerelease version using `--canary` arg that allows prerelease (not necessarily canary)
```
Shouldn't this read `--rc`

JackHowa replied:
added explanation https://github.com/WPMedia/fusion-news-theme-blocks/pull/993/files#r674833117

matthewroach commented on News Theme Development.md:
```diff
@@ -129,6 +129,36 @@ Common phrases needed across multiple blocks are declared in global-phrases-bloc
+
+- Only publish packages to stable tag
+- "release:stable": "lerna publish --dist-tag stable -y", better because no force publishes and only changed ones 
+- ideally should go from: (done https://github.com/JackHowa/lerna-monorepo-components/runs/3126215436?check_suite_focus=true)
```
Not sure I fully understand this reference

JackHowa replied:
yeah that looks like notes; will remove

matthewroach commented on package.json:
```diff
@@ -22,7 +22,11 @@
+    "release:beta": "lerna publish --dist-tag beta --canary --preid beta -y",
+    "release:canary": "lerna publish --dist-tag canary --canary --preid canary -y",
+    "release:rc": "lerna publish --dist-tag rc --canary --preid rc -y",
+    "release:stable": "lerna publish --dist-tag stable minor -y"
```
We should note this will continually update the second semver number so we will be going 5.14 -> 5.15 -> 5.16.
In the future if and when we do correct versioning we would need a way to do major versions when introducing breaking changes, just something to think about

JackHowa replied:
updated and added note

Commit:
fix: Use scss file change as well

Per feedback https://github.com/WPMedia/fusion-news-theme-blocks/pull/993/files#r674591779

Commit:
docs: Remove notes based on testing

Commit:
docs: Update notes for stable release minor versioning

Response to https://github.com/WPMedia/fusion-news-theme-blocks/pull/993/files#r674607132

Commit:
Merge branch 'canary' into TMEDIA-324-lerna-versioning-independent

matthewroach reviewed (approved):
Looks good to me - will need to watch a couple of runs to ensure it's working as fully expected

Commit:
Merge branch 'canary' into TMEDIA-324-lerna-versioning-independent

JackHowa commented:
this should result in no publishes I believe

JackHowa commented:
yes, no publish https://github.com/WPMedia/fusion-news-theme-blocks/actions/workflows/canary-build.yml
